File: src/com/isa/tree/_102_LevelOrder.java
package com.isa.tree;

import java.util.*;

public class _102_LevelOrder {
    public static void main(String[] args) {

    }

    public List<List<Integer>> levelOrder1(TreeNode root) {
        List<List<Integer>> list = new ArrayList();
        if (root == null) return list;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> sub = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                if (node != null) {
                    sub.add(node.val);

                    if (node.left != null) {
                        queue.add(node.left);
                    }
                    if (node.right != null) {
                        queue.add(node.right);
                    }
                }
            }
            list.add(sub);
        }

        return list;
    }


    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> list = new ArrayList();
        if (root == null) return list;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            List<Integer> subList = new ArrayList<>();
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                if (node != null) {
                    subList.add(node.val);

                    if (node.left != null) {
                        queue.offer(node.left);
                    }

                    if (node.right != null) {
                        queue.offer(node.right);
                    }
                }
            }
            list.add(subList);
        }

        return list;
    }
}

File: src/com/isa/tree/_102_levelOrder.kt
package com.isa.tree

import java.util.LinkedList
import java.util.Queue

fun levelOrder(
    root: TreeNode?
): List<List<Int>> {
    val list = mutableListOf<MutableList<Int>>()

    if (root == null) return list

    val queue: Queue<TreeNode> = LinkedList()
    queue.add(root)

    while (!queue.isEmpty()) {
        val subList = mutableListOf<Int>()
        val levelSize = queue.size

        for (i in 0 until levelSize) {
            val node = queue.poll()
            subList.add(node.`val`)

            node.left?.let {
                queue.offer(it)
            }
            node.right?.let {
                queue.offer(it)
            }
        }
        list.add(subList)
    }

    return list
}
